### PR TITLE
Add elastic inversion capability using batch materials

### PR DESCRIPTION
**Pull request by dewenyushu**
Initial demonstration of stress gradient calculation using batch material in the elastic material inversion problem.

### CODE REVIEW

**Comment by dewenyushu** [comment]
poing @zachmprince and @lynnmunday

File: include/userobjects/BatchStressGrad.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+#pragma once
+
+#include "BatchMaterial.h"
+
+typedef BatchMaterial<
+    // tuple representation
+    BatchMaterialUtils::TupleStd,
+    // output data type
+    RankTwoTensor,
```
Comment (hugary1995) [inline include/userobjects/BatchStressGrad.h:18]: So... this implies that p is a scalar-valued property. This is fine for now.

File: src/userobjects/BatchStressGrad.C [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "libmesh/int_range.h"
+
+registerMooseObject("BlackBearApp", BatchStressGrad);
+
+InputParameters
+BatchStressGrad::validParams()
+{
+  auto params = BatchStressGradParent::validParams();
+  params.addRequiredParam<MaterialPropertyName>(
+      "elastic_tensor_derivative", "Name of the elastic tensor derivative material property.");
```
Comment (hugary1995) [inline src/userobjects/BatchStressGrad.C:20]:
```suggestion
      "elasticity_tensor_derivative", "Name of the elasticity tensor derivative material property.");
```
here and a few other places: there's no such a thing called elastic tensor.

File: src/userobjects/BatchStressGrad.C [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                          // here we pass the derivative of elastic tensor wrt to the parameter
+                          "elastic_tensor_derivative",
+                          // here we pass in the forward strain
+                          "forward_mechanical_strain")
+{
+}
+
+void
+BatchStressGrad::batchCompute()
+{
```
Comment (hugary1995) [inline src/userobjects/BatchStressGrad.C:35]: Perhaps a comment here or in the header stating that this is a placeholder until we can swap this out for "actually" batched computation.

File: src/vectorpostprocessors/AdjointStrainStressGradInnerProduct.C [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+{
+  InputParameters params = ElementOptimizationFunctionInnerProduct::validParams();
+  params.addRequiredParam<UserObjectName>(
+      "stress_grad_name",
+      "Name of the stress gradient user object with respect to the material parameter");
+
+  params.addRequiredParam<MaterialPropertyName>(
+      "adjoint_strain_name", "Name of the strain property in the adjoint problem");
+
+  params.addClassDescription("Compute the gradient for elastic material inversion");
```
Comment (hugary1995) [inline src/vectorpostprocessors/AdjointStrainStressGradInnerProduct.C:25]:
```suggestion
  params.addClassDescription("Compute the parameter gradient for linear elastic material inversion");
```

**Review by hugary1995 — CHANGES_REQUESTED** [review]
Cool, thanks!

File: doc/content/source/vectorpostprocessors/AdjointStrainStressGradInnerProduct.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+## Description
+
+This `VectorPostprocessor` computes the gradient of objective function with respect to interested parameter for elastic problem. Specifically,
+\begin{equation}
+  \frac{\text{d}\boldsymbol{F}}{\text{d}p} = - \int \underbrace{(\boldsymbol{L}\lambda)^T}_{\text{adjoint strain}}  \underbrace{\frac{\partial \boldsymbol{C}}{\partial p} (\boldsymbol{L} u)}_{\text{stress gradient}} \text{d}\Omega,
+\end{equation}
+where $\boldsymbol{F}$ is the objective function, $\boldsymbol{L}$ is the differential operator for the elasticity problem, $\lambda$ is the adjoint displacement, $\boldsymbol{C}$ is the elastic tensor, $p$ is the interested parameter, and $u$ is displacement from the forward problem.
+
+Specifically, this object takes the `adjoint strain` and the `stress gradient` as inputs, computes the double contraction of the two inputs, and integrate over the entire domain. Specifically, the `adjoint strain` is the strain from the adjoint problem, the `stress gradient` should be a batch material (e.g., [BatchStressGrad](/BatchStressGrad.md)).
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply doc/content/source/vectorpostprocessors/AdjointStrainStressGradInnerProduct.md:13]: Haha, sorry about the confusion. Yeah I meant the gradient of stress wrt to the interested parameter here but wanted it to be concise. I'll use `gradient of stress w.r.t. parameter` to hopefully avoid further confusion.

File: src/userobjects/BatchStressGrad.C (hunk 3 above)
Comment (dewenyushu) [reply src/userobjects/BatchStressGrad.C:35]: Good point, will do!

**Comment by dewenyushu** [comment]
@hugary1995 Thanks for your review! I guess this indicates a viable path towards using NEML stress derivatives in inverse optimization problems? ;)

If so I'll focus on resolving the civet tests failures and ping you again once I think it is ready

**Comment by peiyichen1995** [comment]
> @hugary1995 Thanks for your review! I guess this indicates a viable path towards using NEML stress derivatives in inverse optimization problems? ;)

Yes, I see that you are getting correct answers in the regression test :)
I'll be working on NEML2 next month.

> If so I'll focus on resolving the civet tests failures and ping you again once I think it is ready

Yes, please.

**Comment by hugary1995** [comment]
> @hugary1995 Thanks for your review! I guess this indicates a viable path towards using NEML stress derivatives in inverse optimization problems? ;)

Yes, I see that you are getting correct answers in the regression test :)
I'll be working on NEML2 next month.

> If so I'll focus on resolving the civet tests failures and ping you again once I think it is ready

Yes, please.

**Comment by dewenyushu** [comment]
This should be ready @hugary1995

File: doc/content/source/vectorpostprocessors/AdjointStrainStressGradInnerProduct.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+\begin{equation}
+  F({u}, \boldsymbol{p}) = ||\boldsymbol{S}{u} - \boldsymbol{u}_m ||^2 + R({u}, \boldsymbol{p}),
+\end{equation}
+where ${u}$ denotes the displacement from the forward problem, $\boldsymbol{S}$ is the sampling operator, $\boldsymbol{u}_m$ are the measurements, $\boldsymbol{p} = [p_1, p_2, ..., p_n]$ is the vector of parameters,  and $R$ is the regularization term.
+
+The gradient of objective function with respect to one scalar parameter ($p_i$) is
+\begin{equation}
+  \frac{\text{d}{F}}{\text{d}p_i} = - \int \underbrace{(\boldsymbol{L}\lambda)^T}_{\text{adjoint strain}}  \underbrace{\frac{\partial \boldsymbol{C}}{\partial p_i} (\boldsymbol{L} u)}_{\text{gradient of stress}} \text{d}\Omega,
+\end{equation}
+where ${F}$ is the objective function, $\boldsymbol{L}$ is the symmetric strain operator for the elasticity problem, $\lambda$ is the adjoint displacement, $\boldsymbol{C}$ is the elasticity tensor, and $p_i$ is the interested scalar parameter.
```
Comment (hugary1995) [inline doc/content/source/vectorpostprocessors/AdjointStrainStressGradInnerProduct.md:19]: What happens to the regularization term? Do we assume it to be zero, or perhaps dR/dp is handled elsewhere?

**Review by hugary1995 — APPROVED** [review]
Looks good to me. I only have one question.

File: include/userobjects/BatchStressGrad.h (hunk 1 above)
Comment (lynnmunday) [reply include/userobjects/BatchStressGrad.h:18]: When is p not scalar valued?

**Comment by lynnmunday** [comment]
@bwspenc or @dschwen Would one of you look at this PR.  This is how we would like to interface with NEML2 for material parameter derivatives for inverse optimization.

File: include/userobjects/BatchStressGrad.h (hunk 1 above)
Comment (dewenyushu) [reply include/userobjects/BatchStressGrad.h:18]: We will need multiple such objects for each p component for that case

File: src/userobjects/BatchStressGrad.C [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "BatchStressGrad.h"
+#include "libmesh/int_range.h"
+
+registerMooseObject("BlackBearApp", BatchStressGrad);
+
+InputParameters
+BatchStressGrad::validParams()
+{
+  auto params = BatchStressGradParent::validParams();
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply src/userobjects/BatchStressGrad.C:18]: Good point. Just did!

File: include/userobjects/BatchStressGrad.h (hunk 1 above)
Comment (hugary1995) [reply include/userobjects/BatchStressGrad.h:18]: ya, as Dewen said we need multiple instances of the object to calculate the derivatives w.r.t. each p, for example we need two objects for the two Lame parameters (more when the stress-strain relation has more parameters). I feel like there will be an overhead in doing this. But I'm not sure how much that overhead will cost us.

File: include/userobjects/BatchStressGrad.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
+//* This file is part of the MOOSE framework
```
Comment (bwspenc) [inline include/userobjects/BatchStressGrad.h:1]: These files will all need the BlackBear header

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Comment by lynnmunday** [comment]
@dewenyushu we need to try moving this to isopod and include blackbear with isopod.

**Comment by dewenyushu** [comment]
Move contributions to https://github.com/idaholab/isopod/pull/78

**Pull request closed without merging**